### ISSUE TITLE
Rule proposal: Detect superfluous requests inside `Promise.all`

### ISSUE THREAD

**Issue by mcmxcdev** [typescript-eslint/typescript-eslint#6631]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

We just encountered a situation where we had a bug where `Promise.all` expected 6 destructured return values, but we called 8 requests in parallel, and it went uncaught.

It would be useful to save unnecessary requests being fired and less code is always good. 

### Fail Cases

```typescript
const [A, B] = await Promise.all([
  requestA,
  requestB,
  requestC,
  requestD,
]);
```

### Pass Cases

```typescript
const [A, B] = await Promise.all([
  requestA,
  requestB,
]);
```

**Comment by JoshuaKGoldberg** [comment]
Oh, very interesting! This technically isn't a case of floating promises though: if `requestC` or `requestD` reject, the surrounding code will still reject. [[TypeScript playground showing the caught rejection (hit _Run_)](https://www.typescriptlang.org/play?#code/IYZwngdgxgBAZgV2gFwJYHsIwLbFRACgEoYBvAKBipikxGRgCcBTARwWfoEEYBeGAAqN02VCGYA6FiHQAbAG7NiAbnKVqtCPSZsO9AEJ9Bw0eKnMAVsyjICAIgDi6ZFAAWwAIR2iq9VWSMYGR+1FSa2gDaXAC6RsAA7ngMQiJiksCysgQRLOycyFwANDp5BtE+IQC+NMAurjAEzIzCjCQUoaHhcpKy6ADm9gDCwAh9rgxNLQBcdsWT6K2qoZXkK+S4+CpAA)]

We _generally_ try not to add special-case logic for built-in APIs that have specialized behaviors such as `Promise.all`. Users have a habit of writing wrappers around those APIs that they then are surprised don't get caught by that logic. My suggestion would be to write a plugin with custom rule(s) that target `Promise.all`, `Promise.allSettled`, and any other API you'd want to make sure nothing gets dropped by.

Thanks for filing though!
